### PR TITLE
Added execution duration metric and stats API for job dispatcher

### PR DESCRIPTION
**Pull request by gmsdelmundo-ttd**
Testing:

- Added unit tests
- While a job was executing, checked http://localhost:9089/metrics and saw the job execution duration in the output

### CODE REVIEW

File: src/main/java/com/uid2/admin/Main.java [hunk 1]
```diff
@@ -161,10 +164,9 @@ public void run() {
 
             //UID2-575 setup a job dispatcher that will write private site data periodically if there is any changes
             //check job for every minute
-            JobDispatcher.getInstance().start(1000 * 60, 3);
             PrivateSiteDataSyncJob job = new PrivateSiteDataSyncJob(config, writeLock);
-            JobDispatcher.getInstance().enqueue(job);
-            JobDispatcher.getInstance().executeNextJob(3);
+            jobDispatcher.enqueue(job);
+            jobDispatcher.executeNextJob();
```
Comment (aulme) [inline src/main/java/com/uid2/admin/Main.java:169]: Why is this changed? Not sure why it did 3 before either though.

File: src/main/java/com/uid2/admin/job/JobDispatcher.java [hunk 2]
```diff
@@ -2,39 +2,56 @@
...
+            int maxRetries,
+            Clock clock) {
+        this.id = id;
+        this.intervalMs = intervalMs;
+        this.maxRetries = maxRetries;
+        this.clock = clock;
+
+        Gauge.builder("uid2.job_dispatcher.execution_duration_ms", this::getExecutionTime)
+                .tag("job_dispatcher", id)
+                .description("gauge for " + id + " execution time")
```
Comment (aulme) [inline src/main/java/com/uid2/admin/job/JobDispatcher.java:45]: Don't need to include "gauge for", just like variable names don't need the word "variable" in them :)

File: src/main/java/com/uid2/admin/job/JobDispatcher.java [hunk 3]
```diff
@@ -141,9 +134,47 @@ private void run(int maxRetries) {
...
+    public boolean isExecutingJob() {
+        return currentJob != null;
+    }
+
+    public boolean isStarted() {
+        return started;
+    }
+
+    private long getExecutionTime() {
+        synchronized (jobLock) {
```
Comment (aulme) [inline src/main/java/com/uid2/admin/job/JobDispatcher.java:176]: I suggest wee don't use lock here - prom will call this very frequently and will keep the lock busy.

File: src/main/java/com/uid2/admin/job/JobDispatcher.java [hunk 4]
```diff
@@ -141,9 +134,47 @@ private void run(int maxRetries) {
...
+
+        synchronized (jobLock) {
+            if (isExecutingJob()) {
+                jobInfos.add(new JobInfo(currentJob, true));
+            }
+            jobInfos.addAll(jobQueue.stream()
+                    .map(job -> new JobInfo(job, false))
+                    .collect(Collectors.toList()));
+        }
+        return jobInfos;
```
Comment (aulme) [inline src/main/java/com/uid2/admin/job/JobDispatcher.java:154]: We need tests for this

File: src/main/java/com/uid2/admin/job/JobDispatcher.java [hunk 5]
```diff
@@ -141,9 +134,47 @@ private void run(int maxRetries) {
...
+        return currentJob != null;
+    }
+
+    public boolean isStarted() {
+        return started;
+    }
+
+    private long getExecutionTime() {
+        synchronized (jobLock) {
+            return currentJob == null ? 0 : ChronoUnit.MILLIS.between(currentJob.getStartedExecutingAt(), Instant.now());
```
Comment (aulme) [inline src/main/java/com/uid2/admin/job/JobDispatcher.java:177]: We need tests for this

File: src/main/java/com/uid2/admin/Main.java (hunk 1 above)
Comment (gmsdelmundo-ttd) [reply src/main/java/com/uid2/admin/Main.java:169]: changing it from singleton to non-singleton

3 was there because you had to pass in `maxRetries`

File: src/main/java/com/uid2/admin/job/JobDispatcher.java (hunk 2 above)
Comment (gmsdelmundo-ttd) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:45]: I think all the other descriptions have the same format so I'm just following those descriptions

File: src/main/java/com/uid2/admin/job/JobDispatcher.java (hunk 3 above)
Comment (gmsdelmundo-ttd) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:176]: Sure will remove it

File: src/main/java/com/uid2/admin/job/JobDispatcher.java (hunk 5 above)
Comment (gmsdelmundo-ttd) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:177]: Actually will update this a bit, will add a relevant test case after

File: src/main/java/com/uid2/admin/job/JobDispatcher.java (hunk 4 above)
Comment (gmsdelmundo-ttd) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:153]: already have tests for this

File: src/main/java/com/uid2/admin/Main.java (hunk 1 above)
Comment (aulme) [reply src/main/java/com/uid2/admin/Main.java:169]: ah right

**Review by aulme — APPROVED** [review]

File: src/main/java/com/uid2/admin/job/JobDispatcher.java [hunk 6]
```diff
@@ -2,39 +2,55 @@
...
+            Clock clock) {
+        this.id = id;
+        this.intervalMs = intervalMs;
+        this.maxRetries = maxRetries;
+        this.clock = clock;
+
+        Gauge.builder("uid2.job_dispatcher.execution_duration_ms", this::getExecutionDuration)
+                .tag("job_dispatcher", id)
+                .description("gauge for " + id + " execution time")
+                .register(Metrics.globalRegistry);
```
Comment (sunnywu) [inline src/main/java/com/uid2/admin/job/JobDispatcher.java:45]: so prometheus will poll for this execution_duration_ms value periodically via the Gauge mechanism? would alert be based on such gauge? (e.g. if prometheus always poll at the time when currnetJob is null then this Gauge wouldn't be helpful? ) 

or should we always store the mnost recent job's executionDuration in JobDispatcher and getExecutionDuration just returns that value? So then regardless when prometheus polls there's always a valid duration value to return for?

**Comment by sunnywu** [comment]
could you also record in MR description what testing you have done so reviewers know the level of testing or give more suggestions if necessary

**Comment by sunnywu** [comment]
how are you planning to setup the alert on "1+ successful job execution over last ~25 mins"? do we need to create a COunter and that get incremented upon every successful job execution and then we can setup the alert based on the rate of the counter increase (over last 25 mins)?

File: src/main/java/com/uid2/admin/job/JobDispatcher.java (hunk 6 above)
Comment (aulme) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:45]: Prom will gather those values periodically, by default I think it's 15sec but we could increase it if we prefer to improve resolution.
A gauge should gather the current value. If our metric is how long the current job has been running and the current job is null, that means currently the job is not running so it indeed has a running time of 0.
Alerting should be fine as long as duration we alert on is greater than scraping period. E.g. if we want to alert on jobs running >1min this should not be a problem.

Comment (sunnywu) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:45]: I think 99% of the time there is no job being run when Prom polls so the gauge will always get 0 which is not informative? and we aren't synchronising `synchronized (jobLock)` inside getExecutionDuration which could create concurrency issue?

So if we use the gauge purely to detect hanged jobs then this is indeed ok tho i thought the gauge can be used as a metric for dashboard too (to see how the duration changes over time, e.g. months as more sites and operators get created). So getting 0 all the time for the gauge isn't useful and probably we need a separate gauge to check for lastSuccessfulRunDuration for us to plot and alert on increase in exec duration?

Comment (gmsdelmundo-ttd) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:45]: Assigning currentJob to a temp variable shouldn't cause any concurrency issue, especially since the function only uses a getter function. However, I've changed it back to a synchronized lock

Comment (gmsdelmundo-ttd) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:45]: And execution duration = 0 can just be interpreted as something is not running. If needed, on the same graph visualising the execution durations, we could also query from the logs when a job actually started executing, so we can get points together with the executionDuration line chart

Comment (sunnywu) [reply src/main/java/com/uid2/admin/job/JobDispatcher.java:45]: as long as the end result is we get an alert if there's no successful job run in last ~25 mins then it's all fine regardless how you implement it. 

so atm we run a job every 20 mins, each time takes around 7-10 seconds so our of an hour (3600 seconds), only 30 out of 3600 seconds the execution duration is not null, so unless Prom polls every second (which I doubt?) otherwise Prom will most likely get 0 most of the time. If it's just for detecting long running/hanged job then it's all good.

**Comment by gmsdelmundo-ttd** [comment]
@sunnywu

> could you also record in MR description what testing you have done so reviewers know the level of testing or give more suggestions if necessary
Sure

how are you planning to setup the alert on "1+ successful job execution over last ~25 mins"? do we need to create a COunter and that get incremented upon every successful job execution and then we can setup the alert based on the rate of the counter increase (over last 25 mins)?
The code change is just to set up the execution duration metrics. The alert to be set up will check if duration > 25 min, and to avoid repeated alerts, maybe just set it to max 3 alerts per hour
